Index: linux-4.14.95/drivers/mmc/host/mtk-mmc/sd.c
===================================================================
--- linux-4.14.95.orig/drivers/mmc/host/mtk-mmc/sd.c
+++ linux-4.14.95/drivers/mmc/host/mtk-mmc/sd.c
@@ -188,7 +188,7 @@ struct msdc_hw msdc0_hw = {
 	.clk_drv        = 4,
 	.cmd_drv        = 4,
 	.dat_drv        = 4,
-	.data_pins      = 4,
+	.data_pins      = 1,
 	.data_offset    = 0,
 	.flags          = MSDC_SYS_SUSPEND | MSDC_CD_PIN_EN | MSDC_REMOVABLE | MSDC_HIGHSPEED,
 //	.flags          = MSDC_SYS_SUSPEND | MSDC_WP_PIN_EN | MSDC_CD_PIN_EN | MSDC_REMOVABLE,
@@ -2325,7 +2325,7 @@ static int msdc_ops_get_cd(struct mmc_ho
 		present = (sdr_read32(MSDC_PS) & MSDC_PS_CDSTS) ? 0 : 1; 
         else
 		present = (sdr_read32(MSDC_PS) & MSDC_PS_CDSTS) ? 1 : 0; 
-   if (host->mmc->caps & MMC_CAP_NEEDS_POLL)
+//ql   if (host->mmc->caps & MMC_CAP_NEEDS_POLL)
        present = 1;
         host->card_inserted = present;  
 #endif        
